### PR TITLE
[serve] Handle cancellation while waiting for system response in router

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

I discovered a small bug in the new routing code when attempting to turn it on by default: the object ref generator was not being cancelled properly if cancellation occurred before the system response was received.

Fixed and added a unit test for this. I've also added some defensive cancellation handling in the router.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/test_utils.py [hunk 1]
```diff
@@ -288,6 +288,8 @@ async def send_signal_on_cancellation(signal_actor: ActorHandle):
     try:
         await asyncio.sleep(100000)
     except asyncio.CancelledError:
+        # Clear the context var to avoid Ray recursively cancelling this method call.
+        ray._raylet.async_task_id.set(None)
```
Comment (edoakes) [inline python/ray/serve/_private/test_utils.py:292]: This is fixing an existing bug in the cancellation tests.

**Review by GeneDer — APPROVED** [review]
LGTM, thanks for catching and fixing this edge case Ed!!

File: python/ray/serve/_private/replica_scheduler/common.py [hunk 2]
```diff
@@ -190,19 +190,19 @@ async def send_request_with_rejection(
         assert (
```
[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/replica_scheduler/common.py:190]: ("this" as the return type of `send_request_with_rejection` function, github doesn't let me comment on it 😅)

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/replica_scheduler/common.py:193]: done! thanks for the catch

**Review by GeneDer — APPROVED** [review]
:shipit:

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -292,7 +292,7 @@ async def _resolve_deployment_responses(
 
     async def schedule_and_send_request(
         self, pr: PendingRequest
-    ) -> Tuple[Union[ray.ObjectRef, "ray._raylet.ObjectRefGenerator"], str]:
+    ) -> Tuple[Union[ray.ObjectRef, ray.ObjectRefGenerator], str]:
```
Comment (zcin) [inline python/ray/serve/_private/router.py:295]: To check my understanding: so `schedule_and_send_request` will always return `ray.ObjectRefGenerator` if we've turned on the feature flag for the new routing? And this is handled by the new `_should_resolve_gen_to_obj_ref`?

File: python/ray/serve/tests/test_cancellation.py [hunk 4]
```diff
@@ -40,7 +40,7 @@ def __init__(self, handle):
 
             @app.get("/")
             async def wait_for_cancellation(self):
-                await self._handle.remote()._to_object_ref()
+                _ = self._handle.remote()
```
Comment (zcin) [inline python/ray/serve/tests/test_cancellation.py:43]: Small nit: why assign it to `_` instead of making the call and not assigning to anything?

**Review by zcin — APPROVED** [review]
LGTM!

File: python/ray/serve/tests/test_cancellation.py (hunk 4 above)
Comment (edoakes) [reply python/ray/serve/tests/test_cancellation.py:43]: hmm I don't remember. I think a linter yelled at me

**Comment by can-anyscale** [comment]
Potentially breaking train and tune tests?

Screenshot 2024-02-21 at 5 17 39 PM

**Comment by zcin** [comment]
> Potentially breaking train and tune tests?

I think it's caused by a power outage trying to fetch https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz:
Screenshot 2024-02-21 at 5 26 09 PM

Seen in failing test logs:
```
Exception: URL fetch failure on https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz: 503 -- Service Unavailable
```

**Comment by can-anyscale** [comment]
Ah good catch thanks @zcin

**Comment by edoakes** [comment]
> > Potentially breaking train and tune tests?
> 
> I think it's caused by a power outage trying to fetch https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz: Screenshot 2024-02-21 at 5 26 09 PM
> 
> Seen in failing test logs:
> 
> ```
> Exception: URL fetch failure on https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz: 503 -- Service Unavailable
> ```

Haha! I thought it was a typo when you said power outage :)

**Pull request merged**